junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/ExtensionValuesStore.java

```diff
@@ -110,7 +110,16 @@ Object remove(Namespace namespace, Object key) {
 
 	private StoredValue getStoredValue(Namespace namespace, Object key) {
 		CompositeKey compositeKey = new CompositeKey(namespace, key);
-		return this.storedValues.get(compositeKey);
+
+		/*
+		 * This method can be called from another ExtensionValuesStore instance,
+		 * so we must synchronized on the monitor.
+		 *
+		 * TODO: Consider rewriting with ConcurrentHashMap.
```

Comment:
We've already discussed this once before in a previous thread, and the conclusion was that a `ConcurrentHashMap` would not help since there are multiple maps involved within a hierarchy of stores.

Do you perceive this differently?